File: SimpleEngine/Engine/GameEngine/Camera.cpp
//
//  Camera.cpp
//  LetsUseSwift
//
//  Created by Sora Sugiyama on 2/15/26.
//

#include "Camera.hpp"
#include <iostream>
Camera::Camera() :
position{0.0f, 1.0f, 5.0f},
unitDeltaFront{1.0f, 0.0f, 0.0f},
unitDeltaSide{0.0f, 0.0f, 1.0f},
unitDeltaUp{0.0f, 1.0f, 0.0f},
yaw{-90.f},
pitch{0.0f}
{
    update();
}

void Camera::moveToCenter()
{
    position = simd::float3{0.0f, 1.0f, 5.0f};
    yaw = -90.0f;
    pitch = 0.0f;
    
    update();
}

void Camera::moveFront(float delta)
{
    position += unitDeltaFront * delta;
}

void Camera::moveSide(float delta)
{
    position += unitDeltaSide * delta;
}

void Camera::moveUp(float delta)
{
    position += unitDeltaUp * delta;
}

void Camera::rotate(float mouseDeltaX, float mouseDeltaY)
{
    yaw += mouseDeltaX;
    pitch += mouseDeltaY;
    
    if (pitch > 89.0f)
    {
        pitch = 89.0f;
    }
    else if (pitch < -89.0f)
    {
        pitch = -89.0f;
    }
    
    update();
}

void Camera::update()
{
    const float cosPitch = cosf(pitch * Consts::PIf / 180.0f);
    simd::float3 newFront;
    newFront.x = cosf(yaw * Consts::PIf / 180.0f) * cosPitch;
    newFront.y = sinf(pitch * Consts::PIf / 180.0f);
    newFront.z = sinf(yaw * Consts::PIf / 180.0f) * cosPitch;
    
    unitDeltaFront = simd::normalize(newFront);
    unitDeltaSide = simd::normalize(simd::cross(unitDeltaFront, unitDeltaUp));
}

simd::float4x4 Camera::getViewMatrix() const
{    
    simd::float3 zAxis = -unitDeltaFront;
    simd::float3 xAxis = simd::normalize(simd::cross(unitDeltaUp, zAxis));
    simd::float3 yAxis = simd::cross(zAxis, xAxis);
    
    return simd::float4x4{
        simd::float4{xAxis.x, yAxis.x, zAxis.x, 0.0f},
        simd::float4{xAxis.y, yAxis.y, zAxis.y, 0.0f},
        simd::float4{xAxis.z, yAxis.z, zAxis.z, 0.0f},
        simd::float4{-simd::dot(xAxis, position), -simd::dot(yAxis, position), -simd::dot(zAxis, position), 1.0f}
    };
}

File: SimpleEngine-RT/SimpleEngine/Camera.swift
//
//  Camera.swift
//  SimpleEngine
//
//  Swift port of the C++ Camera class
//

import simd

final class Camera {
    private var position: SIMD3<Float>
    private var unitDeltaFront: SIMD3<Float>
    private var unitDeltaSide: SIMD3<Float>
    private let unitDeltaUp: SIMD3<Float> = SIMD3(0, 1, 0)
    private var yaw: Float
    private var pitch: Float

    // Cache deg-to-rad conversion to avoid repeated division
    private static let deg2Rad: Float = .pi / 180

    init() {
        position       = SIMD3(0, 1, 5)
        unitDeltaFront = SIMD3(1, 0, 0)
        unitDeltaSide  = SIMD3(0, 0, 1)
        yaw   = -90
        pitch = 0
        update()
    }

    func moveToCenter() {
        position = SIMD3(0, 1, 5)
        yaw   = -90
        pitch = 0
        update()
    }

    func moveFront(_ delta: Float) { position += unitDeltaFront * delta }
    func moveSide (_ delta: Float) { position += unitDeltaSide  * delta }
    func moveUp   (_ delta: Float) { position += unitDeltaUp    * delta }

    func rotate(mouseDeltaX: Float, mouseDeltaY: Float) {
        yaw  += mouseDeltaX
        pitch = max(-89, min(89, pitch + mouseDeltaY))
        update()
    }

    private func update() {
        let cosPitch = cos(pitch * Self.deg2Rad)
        let newFront = SIMD3<Float>(
            cos(yaw * Self.deg2Rad) * cosPitch,
            sin(pitch * Self.deg2Rad),
            sin(yaw * Self.deg2Rad) * cosPitch
        )
        unitDeltaFront = normalize(newFront)
        unitDeltaSide  = normalize(cross(unitDeltaFront, unitDeltaUp))
    }

    var viewMatrix: simd_float4x4 {
        let z = -unitDeltaFront
        let x = normalize(cross(unitDeltaUp, z))
        let y = cross(z, x)
        return simd_float4x4(columns: (
            SIMD4(x.x, y.x, z.x, 0),
            SIMD4(x.y, y.y, z.y, 0),
            SIMD4(x.z, y.z, z.z, 0),
            SIMD4(-dot(x, position), -dot(y, position), -dot(z, position), 1)
        ))
    }
}
